### PR TITLE
fix(prover): speed up LWG and NWG

### PR DESCRIPTION
**Pull request by 0xVolosnikov**
## What ❔

- Loading of proofs for each recursive circuit is now done asynchronously
- Each recursive circuit is now processed in a blocking thread, which speeds up serialization and other CPU-sensitive processing.

Locally, with additional artificial network delays (500 ms), a x5 speed up is observed

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

**Comment by 0xVolosnikov** [comment]
> Did you run tests both ways (sync/async)?

Yes, the difference is noticeable. In different cases, the strategy for using threads by tokio may be different depending on blocking/non-blocking context.

File: prover/crates/bin/witness_generator/src/utils.rs [hunk 1]
```diff
@@ -227,11 +227,15 @@ pub async fn load_proofs_for_job_ids(
...
     object_store: &dyn ObjectStore,
 ) -> Vec<FriProofWrapper> {
-    let mut proofs = Vec::with_capacity(job_ids.len());
+    let mut handles = Vec::with_capacity(job_ids.len());
     for job_id in job_ids {
-        proofs.push(object_store.get(*job_id).await.unwrap());
+        handles.push(object_store.get(*job_id));
     }
-    proofs
+    futures::future::join_all(handles)
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply prover/crates/bin/witness_generator/src/utils.rs:234]: Yes. Do you think this could be a problem?

[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply prover/crates/bin/witness_generator/src/utils.rs:234]: We should have only 32 iterations in this loop

**Comment by 0xVolosnikov** [comment]
> The spawned task may execute on the current thread, or it may be sent to a different thread to be executed. The specifics depend on the current [Runtime](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html) configuration.

https://docs.rs/tokio/latest/tokio/task/fn.spawn.html

**Comment by popzxc** [comment]
> tldr: afaik, async is single-threaded, but blocking tasks aren’t

`tokio::spawn` will run your task on one of the runtime threads (tokio uses the same threadpool to schedule and run stuff). `tokio::spawn_blocking` will run your task on a dedicated blocking threadpool, which is much larger.

`handle.block_on(..)` will block the thread you run on, but it won't block the _runtime_ itself, e.g. runtime will only be utilized for async stuff.
